compiler/src/dmd/opover.d

```diff
@@ -1719,7 +1719,14 @@ private FuncDeclaration findBestOpApplyMatch(Expression ethis, FuncDeclaration f
         if (f.isThis())
         {
             if (!MODimplicitConv(mod, tf.mod))
+            {
+                if (mod & MODFlags.const_)
```

Comment:
The `findBestOpApplyMatch` iterates through all of the overloads of opApply and selects the best match (if there is one) or errors in case you have more than one match. In case there is no match, the function returns null. This patch modifies the code so that if the instance type does not match the opApply instance type and the instance type is `const` then the considered opApply is chosen as the best match. This is wrong for multiple reasons:

- you are simply making `const` a particular case here; what happens with `immutable`, `shared`, etc.?
- you are considering a function that is not a match to be a match; this might break other parts of the compiler that call this function.
- this is not going to work properly when you have multiple opApply's that fail for similar reasons; you are just going to pick the last failing as the best match.
- consider the case when having both a const and a mutable opApply; you will obtain a wrong ambiguity error.

I think that the only solution here is to:

- call `findBestOpApplyMatch`.
-  if there is a single opApply defined and it's not a match, then try calling it anyway to obtain a proper error message
- if you have multiple opApplys and none of them are a match, then the current error message is fine.

So, the fix is not in this function.